scalameta/contrib/src/main/scala/scala/meta/contrib/Extract.scala

```diff
@@ -0,0 +1,21 @@
+package scala.meta.contrib
+
+/**
+  * Logical extraction of B from A.
+  *
+  * Meaning that this is supposed to replicate what the use thinks should happen.
+  * Not the actual class representation
+  *
+  * eg. Extract[Defn.Class, Seq[Stat]]
+  *
+  * is actually extracting the stats from the Template, which is a child of Defn.Class.
+  */
+trait Extract[A, B] {
+  def extract(a: A): B
+}
+
+object Extract {
+  def apply[A, B](f: A => B): Extract[A, B] = new Extract[A, B] {
```

Comment:
What do you think about making the return type `Seq[B]`? That would avoid the need for all the `type Ts = Seq[T]` aliases. It would make it impossible to extract a single `B`, but I think all of the instances in this PR are for some `Seq[B]`.